core/jobs/batch_jobs/manual_voice_artist_names_job.py

```diff
@@ -0,0 +1,458 @@
+# coding: utf-8
+#
+# Copyright 2024 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Jobs used for fetching and saving voice artist names from curated
+exploration models."""
+
+from __future__ import annotations
+
+from core.domain import opportunity_services
+from core.domain import voiceover_services
+from core.jobs import base_jobs
+from core.jobs.io import ndb_io
+from core.jobs.types import job_run_result
+from core.jobs.transforms import job_result_transforms
+from core.jobs.transforms import results_transforms
+from core.platform import models
+
+import apache_beam as beam
+import result
+from typing import Dict, Iterable, List, Tuple
+
+MYPY = False
+if MYPY: # pragma: no cover
+    from mypy_imports import datastore_services
+    from mypy_imports import exp_models
+    from mypy_imports import voiceover_models
+
+datastore_services = models.Registry.import_datastore_services()
+
+(voiceover_models, exp_models) = models.Registry.import_models([
+    models.Names.VOICEOVER, models.Names.EXPLORATION])
+
+
+# TODO(#15613): Here we use MyPy ignore because the incomplete typing of
+# apache_beam library and absences of stubs in Typeshed, forces MyPy to
+# assume that PTransform class is of type Any. Thus to avoid MyPy's error
+# (Class cannot subclass 'PTransform' (has type 'Any')), we added an
+# ignore here.
+class FilterCommitLogModelByExplorationIds(beam.DoFn): # type: ignore[misc]
+    """DoFn to filter exploration commit log entry models based on curated
+    exploration IDs.
+    """
+
+    def process(
+        self,
+        exp_commit_log_model: exp_models.ExplorationCommitLogEntryModel,
+        exploration_ids: List[str]
+    ) -> Iterable[exp_models.ExplorationCommitLogEntryModel]:
+        if exp_commit_log_model.exploration_id in exploration_ids:
+            yield exp_commit_log_model
+
+
+class UpdateVoiceArtistMetadata(beam.DoFn): # type: ignore[misc]
+    """DoFn to update exploration voice artist link and voice artist metdata.
+    """
+    exploration_id_to_voiceover_mapping = {}
```

Comment:
I'm not really sure whether you can rely on these class variables to store data throughout the run -- I'm not quite sure of the lifecycle of a DoFn.

Do other beam jobs do this? If so I guess it's OK, but you might want a second opinion from @JayVivarekar @lkbhitesh07 here. I'm just concerned that multiple threads may mutate this data.

Reply:
The approach is updated.